### PR TITLE
Contract payable addresses

### PR DESCRIPTION
**Pull request by wkaluza**
See https://github.com/fetchai/ledger-api-py/pull/41 for corresponding SDK changes.

Contracts now have addresses of their own, composed as SHA256(concat(owner_address, nonce)). A contract can check its balance and transfer funds using new Etch free functions `balance()` and `transfer(Address, UInt64)`.

Tested using e2e tests.

### CODE REVIEW

File: libs/ledger/include/ledger/chaincode/contract.hpp [hunk 1]
```diff
@@ -113,6 +116,8 @@ class Contract
   Contract &operator=(Contract const &) = delete;
   Contract &operator=(Contract &&) = delete;
 
+  void updateContractContext(ContractContext context);
```
Comment (ejfitzgerald) [inline libs/ledger/include/ledger/chaincode/contract.hpp]: i guess?

```suggestion
  void UpdateContractContext(ContractContext context);
```

File: libs/ledger/include/ledger/upow/synergetic_contract.hpp [hunk 2]
```diff
@@ -84,6 +85,9 @@ class SynergeticContract
   void Attach(StorageInterface &storage);
   void Detach();
 
+  void                   updateContractContext(ContractContext const &context);
```
Comment (ejfitzgerald) [inline libs/ledger/include/ledger/upow/synergetic_contract.hpp]:
```suggestion
  void                   UpdateContractContext(ContractContext const &context);
```

File: libs/ledger/src/chaincode/contract.cpp [hunk 3]
```diff
@@ -229,5 +228,8 @@ Contract::TransactionHandlerMap const &Contract::transaction_handlers() const
   return transaction_handlers_;
 }
 
+Contract::Contract()  = default;
```
Comment (ejfitzgerald) [inline libs/ledger/src/chaincode/contract.cpp]: Haha I guess you know my feelings on these 😁

**Review by ejfitzgerald — CHANGES_REQUESTED** [review]

File: libs/ledger/src/chaincode/contract.cpp (hunk 3 above)
Comment (wkaluza) [reply libs/ledger/src/chaincode/contract.cpp]: This is necessary to keep the unique ptr happy. We should prefer this over using a shared ptr, given we use LTO now.

Comment (wkaluza) [reply libs/ledger/src/chaincode/contract.cpp]: ok, sorted out in conversation - this is only necessary when employing a unique_ptr-to-implementation pattern

File: libs/ledger/include/ledger/upow/synergetic_contract.hpp (hunk 2 above)
Comment (wkaluza) [reply libs/ledger/include/ledger/upow/synergetic_contract.hpp]: Done.

File: libs/ledger/include/ledger/chaincode/contract.hpp (hunk 1 above)
Comment (wkaluza) [reply libs/ledger/include/ledger/chaincode/contract.hpp]: Done.

**Review by ejfitzgerald — APPROVED** [review]
LGTM

**Pull request merged**